<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>LandersLabUtilities Documentation</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            display: flex;
            margin: 0;
            height: 100vh;
        }
        nav {
            width: 200px;
            background-color: #333;
            padding: 20px;
            box-shadow: 2px 0 5px rgba(0,0,0,0.1);
            color: white;
        }
        nav a {
            color: white;
            text-decoration: none;
            display: block;
            margin: 10px 0;
            transition: color 0.3s;
        }
        nav a:hover {
            color: #ffd700;
        }
        main {
            flex-grow: 1;
            padding: 20px;
        }
        h1, h2, h3 {
            color: #333;
        }
        pre {
            background: #2d2d2d;
            color: #ccc;
            padding: 10px;
            border-radius: 5px;
            overflow: auto;
        }
        code {
            color: #ff9100;
        }
        .terminal {
            background: #000;
            color: #0f0;
            padding: 10px;
            border-radius: 5px;
            font-family: monospace;
        }
                table {
            width: 100%;
            border-collapse: collapse;
            margin-bottom: 20px;
        }
        table, th, td {
            border: 1px solid #ddd;
        }
        th, td {
            padding: 8px;
            text-align: left;
        }
        th {
            background-color: #f2f2f2;
        }
    </style>
</head>
<body>
    <nav>
        <h2>Contents</h2>
        <a href="#introduction">Introduction</a>
        <a href="#requirements">Requirements</a>
        <a href="#installation">Installation</a>
        <a href="#functions">Functions</a>
        <ul>
            <li><a href="#genetable">genetable_fromMT</a></li>
            <li><a href="#singlesample">singlesample_fromMT</a></li>
            <li><a href="#jointvcf">jointVCF_fromMT</a></li>
            <li><a href="#extractregions">extract_regions</a></li>
        </ul>
    </nav>
    <main>
        <h1>LandersLabUtilities</h1>
        <p>By Landers Laboratory UMASS Chan Medical School</p>
        <p>Supervisor: John Landers</p>
        <p>Maintainers: Alberto Brusati, Pamela Keagle</p>
        
        <h2 id="introduction">Introduction</h2>
        <p>This documentation provides details about the functions defined in the provided python package (<code>landerslabutilities</code>). These functions, available as a python package on PYPI, were built to lighten the burden of filtering data from the ALS Compute project. In particular, we included gene table extraction, single sample analysis, joint VCF creation, and region extraction. Examples for each function are provided below.</p>

        <h2 id="requirements">Requirements</h2>
        <p>LandersLabUtilities runs on hail clusters or single node. In case you are using our package fro ALS compute project, consider the subsequent configuration:</p>
        <table>
            <tr>
                <th>App</th>
                <th>Features</th>
            </tr>
            <tr>
                <td>Profile CPU</td>
                <td>32</td>
            </tr>
            <tr>
                <td>Profile GB</td>
                <td>120</td>
            </tr>
            <tr>
                <td>Compute type</td>
                <td>Cluster</td>
            </tr>
            <tr>
                <td>Workers</td>
                <td>4</td>
            </tr>
            <tr>
                <td>Workers CPU</td>
                <td>16</td>
            </tr>
        </table>
        <p>If you are filtering samples, first, is mandotory to create a file tab separated with the sampleID in the first column and the subsample in the second (please, use these names for the columns):</p>
        <table>
            <tr>
                <th>sampleID</th>
                <th>subsample</th>
            </tr>
            <tr>
                <td>RES01</td>
                <td>group</td>
            </tr>
            <tr>
                <td>RES02</td>
                <td>group</td>
            </tr>
            <tr>
                <td>RES03</td>
                <td>group</td>
            </tr>
        </table>

        <h2 id="installation">Installation</h2>
        <p><code>Landerslabutilities</code> are completely available on PYPI repository. To easily access the functionality, simply install the package via pip:</p>
        <pre class="terminal">>>> pip3 install landerslabutils</pre>
        <p>And import with the statement:</p>
        <pre class="terminal">>>> import landerslabutils as llu</pre>
        
        <h2 id="functions">Functions</h2>
        
        <h3 id="genetable">1. <code>genetable_fromMT(db=None, phenoFile=None, samples=None, gene=None, variant_type=None, output)</code></h3>
        <p><strong>Description:</strong> Filters a Hail MatrixTable based on gene symbol and consequence type, then exports the results to a TSV file.</p>
        <p><strong>Parameters:</strong></p>
        <ul>
            <li><code>db</code> (str, required): Path to the Hail database (MT) file.</li>
            <li><code>phenoFile</code> (str, required): Path to the phenotype file.</li>
            <li><code>samples</code> (str/list, optional): Subsample your table based on a single sample or list.</li>
            <li><code>gene</code> (str/list, required): Gene symbol(s) to filter on.</li>
            <li><code>variant_type</code> (str/list, optional): Consequence type(s) to filter on.</li>
            <li><code>output</code> (str, required): Path to export the filtered table.</li>
        </ul>
        <p><strong>Example of usage:</strong></p>
        <pre class="terminal">>>> llu.genetable_fromMT(db='/path/to/hail.mt', 
                  phenoFile='/path/to/phenotype_table.tsv', 
                  gene="SOD1",               
                  variant_type=["missense_variant","stop_gained"],  
                  output="/path/to/output_table.tsv") 

>>> llu.genetable_fromMT(db='/path/to/hail.mt', 
                  phenoFile='/path/to/phenotype_table.tsv', 
                  gene=["SOD1","TARDBP"],  
                  variant_type=["missense_variant","stop_gained"],  
                  output="/path/to/output_table.tsv")</pre>
        
        <h3 id="singlesample">2. <code>singlesample_fromMT(db=None, phenoFile=None, sample, gene=None, variant_type=None, output_type="table", output_path, output_name)</code></h3>
        <p><strong>Description:</strong> Filters a Hail MatrixTable based on a single sample and optionally gene and variant type, then exports the results.</p>
        <p><strong>Parameters:</strong></p>
        <ul>
            <li><code>db</code> (str, required): Path to the Hail database (MT) file.</li>
            <li><code>phenoFile</code> (str, required): Path to the phenotype file.</li>
            <li><code>sample</code> (str, required): Sample ID to filter on.</li>
            <li><code>gene</code> (str/list, optional): Gene symbol(s) to filter on.</li>
            <li><code>variant_type</code> (str/list, optional): Consequence type(s) to filter on.</li>
            <li><code>output_type</code> (str, required): "table" or "vcf". Defaults to "table".</li>
            <li><code>output_path</code> (str, required): Path to export the filtered results.</li>
            <li><code>output_name</code> (str, required): Name of the output file.</li>
        </ul>
        <p><strong>Example of usage:</strong></p>
        <pre class="terminal">>>> llu.singlesample_fromMT(db='/path/to/hail.mt', 
                      phenoFile='/path/to/phenotype_table.tsv', 
                      sample="sample", 
                      gene=["TARDBP", "SOD1", "FUS"],  
                      variant_type=None, 
                      output_type="table", 
                      output_path="/path/to/location/", 
                      output_name="sample.table.tsv") 

>>> llu.singlesample_fromMT(db='/path/to/hail.mt', 
                      phenoFile='/path/to/phenotype_table.tsv', 
                      sample="sample", 
                      output_type="vcf", 
                      output_path="/path/to/location/", 
                      output_name="sample.table.vcf.gz")</pre>
        
        <h3 id="jointvcf">3. <code>jointVCF_fromMT(db=None, samples_file=None, output_path=None, output_name=None)</code></h3>
        <p><strong>Description:</strong> Filters a Hail MatrixTable and creates joint VCF files for each chromosome.</p>
        <p><strong>Parameters:</strong></p>
        <ul>
            <li><code>db</code> (str, required): Path to the Hail database (MT) file.</li>
            <li><code>samples_file</code> (str, required): Path to the subsample file.</li>
            <li><code>output_path</code> (str, required): Path to export the VCF files.</li>
            <li><code>output_name</code> (str, required): Name of the output files.</li>
        </ul>
        <p><strong>Example of usage:</strong></p>
        <pre class="terminal">>>> llu.jointVCF_fromMT(db='/path/to/hail.mt', 
               samples_file='/path/to/subsample.txt',  
               output_path='/path/to/location/', 
               output_name='subsample.vcf.gz')</pre>
        
        <h3 id="extractregions">4. <code>extract_regions(db=None, phenoFile=None, samples=None, gene=None, region=None, output_type="table", output=None)</code></h3>
        <p><strong>Description:</strong> Filters a Hail MatrixTable based on genomic regions and optionally gene and variant type, then exports the results.</p>
        <p><strong>Parameters:</strong></p>
        <ul>
            <li><code>db</code> (str, required): Path to the Hail database (MT) file.</li>
            <li><code>phenoFile</code> (str, required): Path to the phenotype file.</li>
            <li><code>samples</code> (str, required): Sample ID(s) to filter on.</li>
            <li><code>gene</code> (str/list, optional): Gene symbol(s) to filter on.</li>
            <li><code>region</code> (list, required): Regions to extract.</li>
            <li><code>output_type</code> (str, required): "table" or "vcf". Defaults to "table".</li>
            <li><code>output</code> (str, required): Path to export the filtered results.</li>
        </ul>
        <p><strong>Example of usage:</strong></p>
        <pre class="terminal">>>> llu.extract_regions(db='/path/to/hail.mt', 
               phenoFile='/path/to/phenotype_table.tsv', 
               samples="sample_id", 
               region=['chr13:32355000-32375000'], 
               output_type="table", 
               output="/path/to/output_name.tsv")</pre>
    </main>
</body>
</html>
